src/main/java/org/apache/sysml/hops/estim/EstimatorBitsetMM.java

```diff
@@ -29,118 +29,116 @@
-		if( !root.getRight().isLeaf() )
-			estim(root.getLeft()); //obtain synopsis
-		BitsetMatrix m1Map = !root.getLeft().isLeaf() ?
-			(BitsetMatrix)root.getLeft().getSynopsis() : new BitsetMatrix(root.getLeft().getData());
-		BitsetMatrix m2Map = !root.getRight().isLeaf() ?
-			(BitsetMatrix)root.getRight().getSynopsis() : new BitsetMatrix(root.getRight().getData());
-		
-		//estimate output density map and sparsity via boolean matrix mult
+		// recursive density map computation of non-leaf nodes
+		if (!root.getLeft().isLeaf())
+			estim(root.getLeft()); // obtain synopsis
+		if (!root.getRight().isLeaf())
+			estim(root.getLeft()); // obtain synopsis
+		BitsetMatrix m1Map = !root.getLeft().isLeaf() ? (BitsetMatrix) root.getLeft().getSynopsis()
+				: new BitsetMatrix(root.getLeft().getData());
+		BitsetMatrix m2Map = !root.getRight().isLeaf() ? (BitsetMatrix) root.getRight().getSynopsis()
+				: new BitsetMatrix(root.getRight().getData());
+
+		// estimate output density map and sparsity via boolean matrix mult
 		BitsetMatrix outMap = m1Map.matMult(m2Map);
-		root.setSynopsis(outMap); //memoize boolean matrix
-		return OptimizerUtils.getSparsity(
-			outMap.getNumRows(), outMap.getNumColumns(), outMap.getNonZeros());
+		root.setSynopsis(outMap); // memoize boolean matrix
+		return OptimizerUtils.getSparsity(outMap.getNumRows(), outMap.getNumColumns(), outMap.getNonZeros());
 	}
 
 	@Override
 	public double estim(MatrixBlock m1, MatrixBlock m2) {
 		BitsetMatrix m1Map = new BitsetMatrix(m1);
 		BitsetMatrix m2Map = new BitsetMatrix(m2);
 		BitsetMatrix outMap = m1Map.matMult(m2Map);
-		return OptimizerUtils.getSparsity( //aggregate output histogram
-			outMap.getNumRows(), outMap.getNumColumns(), outMap.getNonZeros());
+		return OptimizerUtils.getSparsity( // aggregate output histogram
+				outMap.getNumRows(), outMap.getNumColumns(), outMap.getNonZeros());
 	}
 
 	@Override
 	public double estim(MatrixCharacteristics mc1, MatrixCharacteristics mc2) {
 		LOG.warn("Meta-data-only estimates not supported in EstimatorBitsetMM, falling back to EstimatorBasicAvg.");
 		return new EstimatorBasicAvg().estim(mc1, mc2);
 	}
-	
+
 	private static class BitsetMatrix {
 		private final int _rlen;
 		private final int _clen;
 		private long _nonZeros;
 		private BitSet[] _data;
-		
+
 		public BitsetMatrix(int rlen, int clen) {
 			_rlen = rlen;
 			_clen = clen;
 			_data = new BitSet[_rlen];
-			for(int i=0; i<_rlen; i++)
+			for (int i = 0; i < _rlen; i++)
 				_data[i] = new BitSet(_clen);
 			_nonZeros = 0;
 		}
-		
+
 		public BitsetMatrix(MatrixBlock in) {
 			this(in.getNumRows(), in.getNumColumns());
 			init(in);
 		}
-		
+
 		public int getNumRows() {
 			return _rlen;
 		}
-		
+
 		public int getNumColumns() {
 			return _clen;
 		}
-		
+
 		public long getNonZeros() {
 			return _nonZeros;
 		}
-		
+
 		private void init(MatrixBlock in) {
-			if( in.isInSparseFormat() ) {
+			if (in.isInSparseFormat() && in.isEmptyBlock(false)) {
```

Comment:
This would lead to all empty (i.e., not allocated blocks) blocks enter this branch and subsequently fail on accessing the sparse block. Similarly non-empty sparse blocks would enter the dense case and also fail. Please move the check for empty blocks up and directly return in this case.